### ISSUE TITLE
Playing from FileStream using youtube-dl and ffmpeg.

### ISSUE THREAD

**Issue by TybaIt** [filoe/cscore#437]
```csharp
private void ReadCallBack(IAsyncResult asyncResult)
{
    int read = _FFmpegProcess.StandardOutput.BaseStream.EndRead(asyncResult);

    if (read <= 0) {
        _FFmpegProcess.StandardOutput.BaseStream.Close(); 
        return; 
    }

    _bufferingSource.Write(_readBuffer, 0, _readBuffer.Length);

    _FFmpegProcess.StandardOutput.BaseStream.Flush();

    _FFmpegProcess.StandardOutput.BaseStream.BeginRead(_readBuffer, 0, _readBuffer.Length, new AsyncCallback(ReadCallBack), null);
}

public void PlayTrack(string uri, bool isEncounter = false, float volume = 0) {
    if (uri == null || uri.Equals("")) return;

    IWaveSource source;

    if (!FileUtil.IsLocalPath(uri)) {
// Handle youtube urls
        _FFmpegProcess = new Process() {
            StartInfo = new ProcessStartInfo
            {
                FileName = @"C:\Windows\System32\cmd.exe",
                WorkingDirectory = Path.GetDirectoryName(_youtubeDLPath),
                Arguments = $"/C youtube-dl \"{uri}\" -o - | ffmpeg -i pipe:0 -f s16le -ar 44100 -ac 1 pipe:1",
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
            },
            EnableRaisingEvents = true
        };

        _FFmpegProcess.Start();

        _bufferingSource = new WriteableBufferingSource(new WaveFormat(44100, 16, 1)){ FillWithZeros = true };

        _readBuffer = new byte[4096];
        _FFmpegProcess.StandardOutput.BaseStream.BeginRead(_readBuffer, 0, _readBuffer.Length, new AsyncCallback(ReadCallBack), null);

        source = _bufferingSource;

    } else {
// Handle local file paths.
        if (!File.Exists(uri)) return;

        source = CodecFactory.Instance.GetCodec(uri);

    }

    Stop();

    // Setup event for reaching the end of the stream.
    source = new LoopStream(source) { EnableLoop = isEncounter };
    (source as LoopStream).StreamFinished += (o, e) => {
        if (!isEncounter) PlayNext();
    };

    var equalizer = Equalizer.Create10BandEqualizer(source.ToSampleSource());
    var finalSource = equalizer
            .ToStereo()
            .ChangeSampleRate(source.WaveFormat.SampleRate)
            .AppendSource(Equalizer.Create10BandEqualizer, out equalizer)
            .ToWaveSource(source.WaveFormat.BitsPerSample);
            
    _equalizer = equalizer;

    _outputDevice.Initialize(finalSource);
    _initialized = true;

    // Restore previous sound effects.
    ToggleSubmergedFx(_submergedFxEnabled); 

    _outputDevice.Play();
}
```

Hello. I need to write from a filestream,  how do I handle writing to the WriteableBufferingSource?
Any help would be appreciated.

**Comment by TybaIt** [comment]
I was able to successfully start the playback. However, after 2 or 3 seconds it starts speeding up and a lot of cracking occurs. Any clue?

**Comment by filoe** [comment]
Maybe start by writing the raw data you are receiving in the `ReadCallback` to a file and upload it here. 
Also I've spotted a potential issue:
You should change this: 
`_bufferingSource.Write(_readBuffer, 0, _readBuffer.Length);` 
to this:
`_bufferingSource.Write(_readBuffer, 0, read);`

**Comment by TybaIt** [comment]
The file created was 3gb big.  Anyway, I had another approach with threading. Since the WriteableBuffering source doesn't support ``.Position`` I had to use a MemoryStream first. I decided to pre-buffer some data first before I write to the BufferingSource.
I still have the issue that it speeds up and cracks or starts somewhere in the song. This might be due to a race condition now.

```csharp
        public void StartBufferedPlayback() {
            _bufferStream = new MemoryStream();
            new Thread(o => {
                byte[] buffer = new byte[65536]; // 64KB chunks
                int read;
                    
                while (_initialized && ((read = _FFmpegProcess.StandardOutput.BaseStream.Read(buffer, 0, buffer.Length)) > 0)) {
                    var pos = _bufferStream.Position;
                    _bufferStream.Position = _bufferStream.Length;
                    _bufferStream.Write(buffer, 0, read);
                    _bufferStream.Position = pos;
                    _FFmpegProcess.StandardOutput.BaseStream.Flush();
                }
            }).Start();

            
            new Thread(o => {
                // Pre-buffering some data to allow CSCore to start playing
                while (_initialized && _bufferStream.Length < 65536*5)
                    Thread.Sleep(1000);
                while (_initialized) {
                    byte[] buffer = new byte[65536];
                    int read;
                    if ((read = _bufferStream.Read(buffer, 0, buffer.Length)) > 0)
                        _bufferingSource.Write(buffer, 0, read);
                        Thread.Sleep(100);
                }
            }).Start();
        }
````

**Comment by filoe** [comment]
Can you provide a working example project?

**Comment by filoe** [comment]
You might also want to checkout this: https://github.com/filoe/cscore/blob/master/CSCore/Codecs/MP3/Mp3WebStream.cs

It implements a similar streaming mechanism but for mp3 streams.

**Comment by TybaIt** [comment]
I put the class into a quick test project with the ffmpeg.exe and youtube-dl exe
[streamingTest.zip](https://1drv.ms/u/s!AhSrnNZIns1TlBdupjPtuvzkpEDf?e=DrcpPk)

**Comment by filoe** [comment]
The `WriteableBufferingSource` is a ring buffer with a specific bufferSize. You can specify the bufferSize by passing it to the constructor. Otherwise it will be calculated automatically. 
The youtube-file gets downloaded at a certain speed. If the download is faster than the playback consumes data, you will start overwriting the ring buffer before it gets consumed. If that is the case the audio gets damaged and you get the result you were facing. 
You could increase the bufferSize but you would potentially end up in storing 3gb of data within the memory if the download is much faster than the playback. You want to stream it slowly and only add new data if data was already consumed. 

You can rewrite the bufferingProcess like this: 
`public void StartBufferedPlayback() {
            _bufferStream = new MemoryStream();
            new Thread(o => {
                byte[] buffer = new byte[65536]; // 64KB chunks
                int read;
                    
                while (_initialized) {
                    if(_bufferingSource.MaxBufferSize - _bufferingSource.Length > buffer.Length)
                    {
                        read = _FFmpegProcess.StandardOutput.BaseStream.Read(buffer, 0, buffer.Length);
                        if (read <= 0)
                            break;
                        _bufferingSource.Write(buffer, 0, read);
                    }
                    Thread.Sleep(50);
                }
            }).Start();
        }`

It is extremly simplified and you can implement that in a by far better way. But it shows the basic principle. At least it works with this code.

**Comment by TybaIt** [comment]
I thought of the different download speeds and data being consumed too fast but didn't think of the condition with MaxBufferSize. I should have checked the bufferSources fields to realize how I can make my previous preload solution into that dynamic equivalent.   Thanks. I will try that out later when I come back from work. Also. Shouldn't there a Maths.Abs to secure the difference calc in that condition ?

**Comment by TybaIt** [comment]
The bufferingSource doesn't seem to like 2 channels

**Comment by filoe** [comment]
Of course it does. You just have to specify the WaveFormat correctly.

**Comment by TybaIt** [comment]
hmm I did. I made a field at the top to use the same attributes everywhere. (``_channels``) 

```csharp
        private const int _sampleRate = 44100;
        private const int _bits = 16;
        private const int _channels = 1;
        private const int _bitRate = 196; // kbps
        private const string _rawAudioFormat = "s16le";
        private const string _rawAudioCodec = "pcm_s16le";
```
this is what I have currently and it works fine. I played around a bit with every setting and never found a combination that wouldn't make the sound crack with ``_channels = 2``

**Comment by filoe** [comment]
Does ffmpeg return two channels?

**Comment by TybaIt** [comment]
You can force it to with -ac

**Comment by TybaIt** [comment]
ah I think I found the issue. I am converting with ``.ToStereo()`` already. It might be flawed trying to convert existing stereo to stereo.

**Comment by filoe** [comment]
If you convert stereo to stereo it won't have any effect:
https://github.com/filoe/cscore/blob/master/CSCore/FluentExtensions.cs#L101

**Comment by TybaIt** [comment]
okay. I revamped my code quite a bit and I implemented Crossfade. I researched a bit through the past issues and found #80 that helped a lot. I am using the SimpleMixer class and initialize my WasapiOut with it.
However, when trying to initialize the SimpleMixer I get a CSCoreException.
I found that the exception is thrown when the streamrouting has not finished. Furthermore, I found this issue #322 
which says it's my audiodriver. Which cannot be true since my sources worked perfectly fine before the user.
The comment further down below in the issue states that the format is not correct.

**Comment by TybaIt** [comment]
Ok. I made sure the mixer had the same WaveFormat as the source going in. They are equal.

**Comment by TybaIt** [comment]
I am getting closer to a crack free result. The song is completely crack free. But their is still some sort of sub-noise from the bass. Not sure.
I am getting better results changing the order of appending sources.
**From:**
```csharp
            var finalSource = source.ToSampleSource()
                                    .ChangeSampleRate(_sampleRate)
                                    .ToStereo().
                                    .AppendSource(x => new FadeInOut(x){FadeStrategy = new LinearFadeStrategy()}, out _fadeInOut)
                                    .AppendSource(Equalizer.Create10BandEqualizer, out _equalizer);```
**To:**
```csharp
            var finalSource = source.ToSampleSource()
                                    .AppendSource(x => new FadeInOut(x){FadeStrategy = new LinearFadeStrategy()}, out _fadeInOut)
                                    .AppendSource(Equalizer.Create10BandEqualizer, out _equalizer)
                                    .ChangeSampleRate(_sampleRate)
                                    .ToStereo();```

**Comment by TybaIt** [comment]
I am at a point where I get almost plop free results.
My last two issues:
 - when fading with LinearFadeStrategy there is always a very short plopping before the fading finishes and a source disposes.
 - the first SampleSourced added to the mixer will have extreme cracking. Subsequent are played almost cleanly. I assume this has something todo with the buffer in SimpleMixer not being ready on the first source.

**Comment by TybaIt** [comment]
@filoe

**Comment by TybaIt** [comment]
The issue has been that the ``LinearFadeStrategy`` pulled Channel and SampleRate from the wrong waveformat not the source of the audio signal. I had to assign these attributes manually and now everything works perfectly.
However, I removed the mixer in favor of a *replace & dispose* pattern for the soundOut (WasApiOut). The advantage of this is that crossfade works with different samplerates and channels which I think is more appropriate when mixing local files with unknown audio streams.
This issue can be closed if you don't have anything to remark on. Thanks for the help.
